pkg/collector/smi_collector.go

```diff
@@ -0,0 +1,109 @@
+package collector
+
+import (
+	"log"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/aks-periscope/pkg/interfaces"
+	"github.com/Azure/aks-periscope/pkg/utils"
+)
+
+// SmiCollector defines an Smi Collector struct
+type SmiCollector struct {
+	BaseCollector
+}
+
+var _ interfaces.Collector = &SmiCollector{}
+
+// NewSmiCollector is a constructor
+func NewSmiCollector(exporter interfaces.Exporter) *SmiCollector {
+	return &SmiCollector{
+		BaseCollector: BaseCollector{
+			collectorType: Smi,
+			exporter:      exporter,
+		},
+	}
+}
+
+// Collect implements the interface method
+func (collector *SmiCollector) Collect() error {
+	// Get all CustomResourceDefinitions in the cluster
+	allCrdsList, err := utils.GetResourceList([]string{"get", "crds", "-o", "jsonpath={..metadata.name}"}, " ")
+	if err != nil {
+		return err
+	}
+
+	// Directory where logs will be written to
+	rootPath, err := utils.CreateCollectorDir(collector.GetName())
+	if err != nil {
+		return err
+	}
+
+	// Filter to obtain a list of Smi CustomResourceDefinitions in the cluster
+	crdNameContainsSmiPredicate := func(s string) bool { return strings.Contains(s, "smi-spec.io") }
+	smiCrdsList := filter(allCrdsList, crdNameContainsSmiPredicate)
+	if len(smiCrdsList) == 0 {
+		// TODO should this return an error?
+		log.Printf("Cluster does not contain any SMI CustomResourceDefinitions\n")
+		return nil
+	}
+
+	collectSmiCrdDefinitions(collector, filepath.Join(rootPath, "smi_crd_definitions"), smiCrdsList)
+	collectSmiCustomResourcesFromAllNamespaces(collector, filepath.Join(rootPath, "smi_custom_resources"), smiCrdsList)
+
+	return nil
+}
+
+func collectSmiCrdDefinitions(collector *SmiCollector, rootPath string, smiCrdsList []string) {
+	for _, smiCrd := range smiCrdsList {
+		fileName := smiCrd + "_definition.yaml"
+		kubeCmd := []string{"get", "crd", smiCrd, "-o", "yaml"}
+		if err := collector.CollectKubectlOutputToCollectorFiles(rootPath, fileName, kubeCmd); err != nil {
+			log.Printf("Failed to collect %s: %+v", fileName, err)
```

Comment:
💡 is this `fmt.Errorf(` or `log.Printf` I will let you decide, but the message `Failed` suggest its `Error`?

Reply:
This is resolved in 5b56484f608330da011cacf14314c75cf620ddad.
This should not return an error and should not give up trying to find the YAML definitions of the other SMI CRDs.
When periscope fails to obtain the YAML definition of a single SMI CRD, it should log a message noting that it skipped collecting the YAML of *only this* SMI CRD (should not error). Periscope should then proceed with collecting the YAML definitions of other SMI CRDs.

```go
	for _, smiCrd := range smiCrdsList {
		fileName := smiCrd + "_definition.yaml"
		kubeCmd := []string{"get", "crd", smiCrd, "-o", "yaml"}
		if err := collector.CollectKubectlOutputToCollectorFiles(rootPath, fileName, kubeCmd); err != nil {
			log.Printf("Skipping: unable to collect yaml definition of %s to %s: %+v", smiCrd, fileName, err)
                         ^^^ Do not give up and error - continue collecting the yaml definitions of the other SMI CRDs.
		}
	}
```